### PR TITLE
Adding DevTest Lab ARM Template facts

### PR DESCRIPTION
**Pull request by zikalino**
##### SUMMARY
Adding DTL ARM Template facts module.

##### ISSUE TYPE
- New Module Pull Request

##### COMPONENT NAME
azure_rm_devtestlabarmtemplate_facts

### CODE REVIEW

**Comment by Fred-sun** [comment]
@yungezz  @yuwzho

File: lib/ansible/modules/cloud/azure/azure_rm_devtestlabarmtemplate_facts.py [hunk 1]
```diff
@@ -0,0 +1,217 @@
+#!/usr/bin/python
+#
+# Copyright (c) 2018 Zim Kalinowski, <zikalino@microsoft.com>
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_devtestlabarmtemplate_facts.py]: 2019

File: lib/ansible/modules/cloud/azure/azure_rm_devtestlabarmtemplate_facts.py [hunk 2]
```diff
@@ -0,0 +1,217 @@
...
+                    'status': ['preview'],
+                    'supported_by': 'community'}
+
+
+DOCUMENTATION = '''
+---
+module: azure_rm_devtestlabarmtemplate_facts
+version_added: "2.8"
+short_description: Get Azure Arm Template facts.
+description:
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_devtestlabarmtemplate_facts.py:21]: arm to full name?

File: lib/ansible/modules/cloud/azure/azure_rm_devtestlabarmtemplate_facts.py [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+        description:
+            - The name of the lab.
+        required: True
+    artifact_source_name:
+        description:
+            - The name of the artifact source.
+        required: True
+    name:
+        description:
+            - The name of the azure Resource Manager template.
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_devtestlabarmtemplate_facts.py]: Azure?

File: lib/ansible/modules/cloud/azure/azure_rm_devtestlabarmtemplate_facts.py [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+try:
+    from msrestazure.azure_exceptions import CloudError
+    from azure.mgmt.devtestlabs import DevTestLabsClient
+    from msrest.serialization import Model
+except ImportError:
+    # This is handled in azure_rm_common
+    pass
+
+
+class AzureRMArmTemplateFacts(AzureRMModuleBase):
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_devtestlabarmtemplate_facts.py]: AzureDevTestLabArmTemplateFacts?

Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_devtestlabarmtemplate_facts.py]: yes, fixed

File: lib/ansible/modules/cloud/azure/azure_rm_devtestlabarmtemplate_facts.py [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+options:
+    resource_group:
+        description:
+            - The name of the resource group.
+        required: True
+    lab_name:
+        description:
+            - The name of the lab.
+        required: True
+    artifact_source_name:
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_devtestlabarmtemplate_facts.py:33]: why it need artifact_source_name? is this copied from artifact source facts module?

**Review by yungezz — COMMENTED** [review]
code itself looks good. Just didn't get difference betwen armtemplate_facts and artifactsource_facts. can these 2 combine  as one?

File: lib/ansible/modules/cloud/azure/azure_rm_devtestlabarmtemplate_facts.py (hunk 5 above)
Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_devtestlabarmtemplate_facts.py:33]: it's subresourcece of artifact_source.
Artifacts source is a GitHub/VSO repository, which can contain:
- artifacts
- ARM templates
so these two are subresources of artifacts source.

Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_devtestlabarmtemplate_facts.py:33]: i return artifacts_source_name just for consistency.
somebody may query ARM templates from 2 artifact sources and then combine the lists for instance.

File: lib/ansible/modules/cloud/azure/azure_rm_devtestlabarmtemplate_facts.py (hunk 1 above)
Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_devtestlabarmtemplate_facts.py]: fixed already

File: lib/ansible/modules/cloud/azure/azure_rm_devtestlabarmtemplate_facts.py (hunk 2 above)
Comment (zikalino) [reply lib/ansible/modules/cloud/azure/azure_rm_devtestlabarmtemplate_facts.py:21]: updated to "Azure DevTest Lab ARM Template"

File: lib/ansible/modules/cloud/azure/azure_rm_devtestlabarmtemplate_facts.py [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+
+    def list(self):
+        response = None
+        results = []
+        try:
+            response = self.mgmt_client.arm_templates.list(resource_group_name=self.resource_group,
+                                                           lab_name=self.lab_name,
+                                                           artifact_source_name=self.artifact_source_name)
+            self.log("Response : {0}".format(response))
+        except CloudError as e:
```
Comment (yungezz) [inline lib/ansible/modules/cloud/azure/azure_rm_devtestlabarmtemplate_facts.py:171]: should we fail when list get null?

**Review by yungezz — APPROVED** [review]

**Pull request merged**